File: gametest/common/src/main/scripts/oc2test/suites/guest_libraries/test_robot.py
import sys

from harness import expect, report

state = {
    "clock": 0,
    "events": [],
    "results": [],
    "queue_replies": [],
    "result_calls": 0,
    "wait_calls": 0,
    "queue_calls": 0,
    "last_event_type": None,
    "last_queued": None,
}


class FakeTime:
    def ticks_ms(self):
        return state["clock"]

    def ticks_add(self, ticks, delta):
        return ticks + delta

    def ticks_diff(self, a, b):
        return a - b


sys.modules["time"] = FakeTime()


def pop(key, default=None):
    queue = state[key]
    return queue.pop(0) if queue else default


class FakeDevice:
    def getLastActionId(self):
        return 1

    def getActionResult(self, action_id):
        state["result_calls"] += 1
        return pop("results")

    def move(self, direction):
        return self._queue("move")

    def turn(self, direction):
        return self._queue("turn")

    def _queue(self, kind):
        state["queue_calls"] += 1
        state["last_queued"] = kind
        return pop("queue_replies", True)


class FakeBus:
    def find(self, type_name):
        return FakeDevice() if type_name == "robot" else None

    def wait_event(self, timeout=None, event_type=None):
        state["wait_calls"] += 1
        state["last_event_type"] = event_type
        while True:
            event = pop("events")
            if event is None:
                state["clock"] += timeout  # the wait timed out
                return None
            if event_type is None or event.get("type") == event_type:
                return event


class _Stub:
    pass


fake_devices = _Stub()
sys.modules["devices"] = fake_devices


def reset():
    state["clock"] = 0
    state["events"] = []
    state["results"] = []
    state["queue_replies"] = []
    state["result_calls"] = 0
    state["wait_calls"] = 0
    state["queue_calls"] = 0
    state["last_event_type"] = None
    state["last_queued"] = None
    fake_devices.bus = FakeBus()
    if "robot" in sys.modules:
        del sys.modules["robot"]
    return __import__("robot")


def completed(action_id, result):
    state["events"].append({"type": "robotActionCompleted",
                            "data": {"actionId": action_id, "result": result}})


# Completion event
robot = reset()
state["results"] = ["INCOMPLETE"]
completed(1, "SUCCESS")
expect("move succeeded on the event", robot.move("forward"), True)
expect("only the initial poll was needed", state["result_calls"], 1)
expect("the wait blocked on the event channel", state["wait_calls"], 1)
expect("it asked for the completion event by name",
       state["last_event_type"], "robotActionCompleted")
expect("it queued a move", state["last_queued"], "move")

# Failure event
robot = reset()
state["results"] = ["INCOMPLETE"]
completed(1, "FAILURE")
expect("failure is reported as failure", robot.move("forward"), False)
expect("failure needed no extra poll", state["result_calls"], 1)

# Another action's event
robot = reset()
state["results"] = ["INCOMPLETE", "INCOMPLETE", "SUCCESS"]
completed(99, "FAILURE")
expect("another action's event is ignored", robot.move("forward"), True)
expect("we polled again after the stray event", state["result_calls"], 3)

# Another event type
robot = reset()
state["results"] = ["INCOMPLETE"]
state["events"] = [{"type": "devicesChanged", "gen": 3}]
completed(1, "SUCCESS")
expect("a devicesChanged did not end the wait", robot.move("forward"), True)
expect("it still ended on the completion, with no extra poll", state["result_calls"], 1)

# Timeout falls back to polling
robot = reset()
state["results"] = ["INCOMPLETE", "INCOMPLETE", "SUCCESS"]
expect("polling still completes", robot.move("forward"), True)
expect("one wait per poll", state["wait_calls"], 2)
expect("each timed-out wait cost its full interval", state["clock"], 2000)

# Action timeout
robot = reset()
state["results"] = ["INCOMPLETE"] * 100
expect("a stuck action times out", robot.move("forward", 3000), False)
expect("it gave up within the timeout", state["clock"] <= 4000, True)

# Unknown action id
robot = reset()
expect("an unknown action is not success", robot.move("forward"), False)

# Queue full until a slot frees
robot = reset()
state["queue_replies"] = [False, False, True]
state["results"] = ["SUCCESS"]
completed(1, "SUCCESS")
expect("queueing retried until it fit", robot.move_async("forward"), True)
expect("one wait per rejected attempt", state["wait_calls"], 2)
expect("the queue wait also names the event",
       state["last_event_type"], "robotActionCompleted")
expect("it kept trying until the queue took it", state["queue_calls"], 3)

# Queue timeout
robot = reset()
state["queue_replies"] = [False] * 100
expect("a permanently full queue times out", robot.move_async("forward", 3000), False)

# Turning
robot = reset()
state["results"] = ["INCOMPLETE"]
completed(1, "SUCCESS")
expect("turn waits the same way", robot.turn("left"), True)
expect("turn used the event", state["result_calls"], 1)
expect("turn queued a turn, not a move", state["last_queued"], "turn")

report()

File: gametest/common/src/main/scripts/oc2test/suites/guest_libraries/test_robot.lua
local harness = require("harness")
local expect, report = harness.expect, harness.report

local state = {
  clock = 0,
  events = {},
  results = {},
  queueReplies = {},
  resultCalls = 0,
  waitCalls = 0,
  queueCalls = 0,
  lastEventType = nil,
  lastQueued = nil,
  statusColor = 0xFFFFFF,
  statusValue = 1,
  statusGets = 0,
  statusSets = 0,
}

package.preload["oc2.clock"] = function()
  local fake = {}
  function fake.ms() return state.clock end
  function fake.deadline(timeout) return state.clock + timeout end
  function fake.remaining(deadline)
    local left = deadline - state.clock
    return left > 0 and left or 0
  end
  function fake.expired(deadline) return deadline - state.clock <= 0 end
  return fake
end

local function pop(key)
  return table.remove(state[key], 1)
end

local function queueAction(kind)
  return function()
    state.queueCalls = state.queueCalls + 1
    state.lastQueued = kind
    local reply = pop("queueReplies")
    return reply == nil or reply
  end
end

local device = {
  getLastActionId = function() return 1 end,
  getActionResult = function()
    state.resultCalls = state.resultCalls + 1
    return pop("results")
  end,
  move = queueAction("move"),
  turn = queueAction("turn"),
  detect = function(_, side)
    state.lastDetectSide = side
    return "solid"
  end,
  getStatusColor = function()
    state.statusGets = state.statusGets + 1
    return state.statusColor
  end,
  setStatusColor = function(_, value)
    state.statusSets = state.statusSets + 1
    state.statusColor = value & 0xFFFFFF
    return state.statusColor
  end,
  getStatusValue = function()
    state.statusGets = state.statusGets + 1
    return state.statusValue
  end,
  setStatusValue = function(_, value)
    state.statusSets = state.statusSets + 1
    state.statusValue = math.min(math.max(value, 0), 1)
    return state.statusValue
  end,
}

package.preload["devices"] = function()
  return {
    find = function(_, typeName) return typeName == "robot" and device or nil end,
    waitEvent = function(_, timeout, eventType)
      state.waitCalls = state.waitCalls + 1
      state.lastEventType = eventType
      while true do
        local event = pop("events")
        if not event then
          state.clock = state.clock + timeout -- the wait timed out
          return nil
        end
        if not eventType or event.type == eventType then
          return event
        end
      end
    end,
  }
end

local function reset()
  state.clock = 0
  state.events = {}
  state.results = {}
  state.queueReplies = {}
  state.resultCalls = 0
  state.waitCalls = 0
  state.queueCalls = 0
  state.lastEventType = nil
  state.lastQueued = nil
  state.statusColor = 0xFFFFFF
  state.statusValue = 1
  state.statusGets = 0
  state.statusSets = 0
  package.loaded["robot"] = nil
  return require("robot")
end

local function completed(actionId, result)
  state.events[#state.events + 1] =
      { type = "robotActionCompleted", data = { actionId = actionId, result = result } }
end

-- Completion event
local robot = reset()
state.results = { "INCOMPLETE" }
completed(1, "SUCCESS")
expect("move succeeded on the event", robot.move("forward"), true)
expect("only the initial poll was needed", state.resultCalls, 1)
expect("the wait blocked on the event channel", state.waitCalls, 1)
expect("it asked for the completion event by name", state.lastEventType, "robotActionCompleted")
expect("it queued a move", state.lastQueued, "move")

-- Failure event
robot = reset()
state.results = { "INCOMPLETE" }
completed(1, "FAILURE")
expect("failure is reported as failure", robot.move("forward"), false)
expect("failure needed no extra poll", state.resultCalls, 1)

-- Another action's event
robot = reset()
state.results = { "INCOMPLETE", "INCOMPLETE", "SUCCESS" }
completed(99, "FAILURE")
expect("another action's event is ignored", robot.move("forward"), true)
expect("we polled again after the stray event", state.resultCalls, 3)

-- Another event type
robot = reset()
state.results = { "INCOMPLETE" }
state.events = { { type = "devicesChanged", gen = 3 } }
completed(1, "SUCCESS")
expect("a devicesChanged did not end the wait", robot.move("forward"), true)
expect("it still ended on the completion, with no extra poll", state.resultCalls, 1)

-- Timeout falls back to polling
robot = reset()
state.results = { "INCOMPLETE", "INCOMPLETE", "SUCCESS" }
expect("polling still completes", robot.move("forward"), true)
expect("one wait per poll", state.waitCalls, 2)
expect("each timed-out wait cost its full interval", state.clock, 2000)

-- Action timeout
robot = reset()
for i = 1, 100 do state.results[i] = "INCOMPLETE" end
expect("a stuck action times out", robot.move("forward", 3000), false)
expect("it gave up within the timeout", state.clock <= 4000, true)

-- Unknown action id
robot = reset()
expect("an unknown action is not success", robot.move("forward"), false)

-- Queue full until a slot frees
robot = reset()
state.queueReplies = { false, false, true }
state.results = { "SUCCESS" }
completed(1, "SUCCESS")
expect("queueing retried until it fit", robot.moveAsync("forward"), true)
expect("one wait per rejected attempt", state.waitCalls, 2)
expect("the queue wait also names the event", state.lastEventType, "robotActionCompleted")
expect("it kept trying until the queue took it", state.queueCalls, 3)

-- Queue timeout
robot = reset()
for i = 1, 100 do state.queueReplies[i] = false end
expect("a permanently full queue times out", robot.moveAsync("forward", 3000), false)

-- Turning
robot = reset()
state.results = { "INCOMPLETE" }
completed(1, "SUCCESS")
expect("turn waits the same way", robot.turn("left"), true)
expect("turn used the event", state.resultCalls, 1)
expect("turn queued a turn, not a move", state.lastQueued, "turn")

-- Detecting
robot = reset()
expect("detect passes the side through", robot.detect(robot.side.front), "solid")
expect("and sends the side it was given", state.lastDetectSide, "front")
expect("detect is a query, not a queued action", state.queueCalls, 0)
expect("calling detect without a side is an error",
       select(1, pcall(robot.detect)), false)

-- Status light
robot = reset()
expect("the color reads back", robot.statusColor(), 0xFFFFFF)
expect("reading is one call", state.statusGets, 1)
expect("setting answers with what was applied", robot.statusColor(0x123456), 0x123456)
expect("setting went through the setter", state.statusSets, 1)
expect("setting did not also read", state.statusGets, 1)

robot = reset()
expect("the fill reads back", robot.statusValue(), 1)
expect("a zero fill is a set, not a read", robot.statusValue(0), 0)
expect("zero went through the setter", state.statusSets, 1)
expect("zero did not also read", state.statusGets, 1)

report()
